tests/base.py

```diff
@@ -0,0 +1,2602 @@
+                data = cursor.fetchall()
+                utils.memoryview_rows_to_bytes(data)
+                return data
+
+    def execute_prepared_statement(self, query, args=None):
+        if args:
+            self.fail("<args> param for executor {} not supported now".format(self.__class__))
+        kwargs = get_connect_args(self.connection_args.port)
+        with psycopg2.connect(
+                host=self.connection_args.host,
+                dbname=self.connection_args.dbname, **kwargs) as connection:
+            with connection.cursor(
+                    cursor_factory=psycopg2.extras.DictCursor) as cursor:
+                cursor.execute("prepare test_statement as {}".format(query))
+                cursor.execute("execute test_statement")
+                data = cursor.fetchall()
+                utils.memoryview_rows_to_bytes(data)
+                return data
+
+
+# Place arguments in a linear order
+# For example, if query is "$2 $1 $1 $3" and arguments are [a, b, c]
+# The resulted query would be "%s %s %s %s" and arguments would be [b a a c]
+def replace_pg_placeholders_with_psycopg3(query, args):
+    indexes = re.findall(r'\$([0-9]+)', query)
+    query = re.sub(r'\$[0-9]+', '%s', query)
+    # minus 1 because python uses 0-based array, while postgres 1-based
+    newargs = [args[int(i) - 1] for i in indexes]
+    return query, newargs
+
+
+class Psycopg3Executor(QueryExecutor):
+    def _execute(self, query, args=None, prepare=None):
+        query, args = replace_pg_placeholders_with_psycopg3(query, args)
+
+        connection_args = get_connect_args(self.connection_args.port)
+
+        connection_args['sslrootcert'] = self.connection_args.ssl_ca
+        connection_args['sslkey'] = self.connection_args.ssl_key
+        connection_args['sslcert'] = self.connection_args.ssl_cert
+        connection_args['host'] = self.connection_args.host
+        connection_args['dbname'] = self.connection_args.dbname
+
+        with psycopg3.connect(**connection_args) as conn:
+            with conn.cursor(
+                    row_factory=psycopg3.rows.dict_row) as cursor:
+                cursor.execute(query, args, prepare=prepare)
+                try:
+                    data = cursor.fetchall()
+                except psycopg3.ProgrammingError as ex:
+                    # psycopg3 throws an error if we want to fetch rows on
+                    # responseless query, like insert. So ignore it.
+                    if str(ex) == "the last operation didn't produce a result":
+                        return []
+                    else:
+                        raise
+                utils.memoryview_rows_to_bytes(data)
+                return data
+
+    def execute(self, query, args=None):
+        return self._execute(query, args, prepare=False)
+
+    def execute_prepared_statement(self, query, args=None):
+        return self._execute(query, args, prepare=True)
+
+
+class ExecutorMixin:
```

Comment:
what do you think about leaving here all base staff with global variable declarations and commonly used functions and all mixins with tests move to separate module like `test_common.py`? To not have here any runnable test cases, only declarations and common function? On the first view it looks like will simplify reading by separation runnable testcases and neighbour mixins and other not-runnable stuff. Plus, unittest has convention that all runnable modules should be named `test*`, so we will fit in this convention too, if base.py will not have it.

plus, we can some of common functions move to `utils.py` which not depends on global variables.